src/fps/proposal/MultisigProposal.sol

```diff
@@ -0,0 +1,854 @@
+        IGnosisSafe safe = IGnosisSafe(multisig);
+        startingOwners = safe.getOwners();
+
+        for (uint256 i = 0; i < config.allowedStorageWriteAccesses.length; i++) {
+            for (uint256 j = 0; j < chains.length; j++) {
+                _allowedStorageAccesses.add(
+                    addresses.getAddress(config.allowedStorageWriteAccesses[i], chains[j].chainId)
+                );
+            }
+        }
+    }
+
+    /// @notice function to be used by forge script.
+    /// @dev use flags to determine which actions to take
+    ///      this function shoudn't be overriden.
+    function _processProposal() internal override {
+        mock();
+        build();
+        simulate();
+        validate();
+        print();
+    }
+
+    /// @notice get the calldata to be executed by safe
+    /// @return data The calldata to be executed
+    function getCalldata() public view override returns (bytes memory data) {
+        /// get proposal actions
+        (address[] memory targets, uint256[] memory values, bytes[] memory arguments) = getProposalActions();
+
+        /// create calls array with targets and arguments
+        Call3Value[] memory calls = new Call3Value[](targets.length);
+
+        for (uint256 i; i < calls.length; i++) {
+            require(targets[i] != address(0), "Invalid target for multisig");
+            calls[i] = Call3Value({target: targets[i], allowFailure: false, value: values[i], callData: arguments[i]});
+        }
+
+        /// generate calldata
+        data = abi.encodeWithSignature("aggregate3Value((address,bool,uint256,bytes)[])", calls);
+    }
+
+    /// @notice print the data to sig by EOA for single multisig
+    function printDataToSign() public view {
+        console.logBytes(_getDataToSign(multisig, getCalldata()));
+    }
+
+    /// @notice print the hash to approve by EOA for single multisig
+    function printHashToApprove() public view {
+        bytes32 hash = keccak256(_getDataToSign(multisig, getCalldata()));
+        console.logBytes32(hash);
+    }
+
+    /// @notice get the data to sign by EOA for single multisig
+    /// @param safe The address of the safe
+    /// @param data The calldata to be executed
+    /// @return The data to sign
+    function _getDataToSign(address safe, bytes memory data) internal view returns (bytes memory) {
+        return IGnosisSafe(safe).encodeTransactionData({
+            to: MULTICALL3_ADDRESS,
+            value: 0,
+            data: data,
+            operation: Enum.Operation.DelegateCall,
+            safeTxGas: 0,
+            baseGas: 0,
+            gasPrice: 0,
+            gasToken: address(0),
+            refundReceiver: address(0),
+            _nonce: nonce
+        });
+    }
+
+    /// @notice simulate the proposal by approving from owners and then executing
+    function simulate() public override {
+        bytes memory data = getCalldata();
+        bytes32 hash = keccak256(_getDataToSign(multisig, data));
```

Comment:
This logic happens a bunch of times in this file, is there an opportunity to refactor this into a function?

Reply:
Ah, I was thinking more about this being reused in multiple places:
```
bytes32 hash = keccak256(_getDataToSign(multisig, getCalldata()))
```

A useful function might be:

```solidity
    function getHash() internal view returns (bytes32, bytes memory) {
        bytes memory data = getCalldata();
        bytes32 hash = keccak256(_getDataToSign(multisig, data));
        return (hash, data);
    }
```

Reply:
Awesome, resolving.